thentos-core/src/Thentos/Frontend/CSRF.hs

```diff
@@ -0,0 +1,99 @@
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE PackageImports        #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeOperators         #-}
+module Thentos.Frontend.CSRF where
+
+import "cryptonite" Crypto.Hash (SHA256)
+import "cryptonite" Crypto.MAC.HMAC (HMAC,hmac)
+import "cryptonite" Crypto.Random (MonadRandom(getRandomBytes))
+import Data.Configifier ((>>.))
+import Data.ByteArray.Encoding (convertToBase, convertFromBase, Base(Base16))
+
+import qualified Data.ByteString as SBS
+
+import Thentos.Prelude hiding (MonadRandom)
+import Thentos.Frontend.Types
+import Thentos.Types (ThentosSessionToken(fromThentosSessionToken))
+
+import qualified Thentos.Action.Unsafe as U
+
+newtype CsrfSecret = CsrfSecret SBS
+    deriving (Show)
+newtype CsrfNonce  = CsrfNonce  SBS
+
+-- | This ONLY checks the format of a given CSRF secret not if it has been randomly choosen (duh!).
+validFormatCsrfSecretField :: Maybe ST -> Bool
+validFormatCsrfSecretField ms
+    | Just t <- ms
+    , Right s' <- convertFromBase Base16 (cs t :: SBS)
+    = SBS.length s' == 32
+    | otherwise = False
+
+-- | This ONLY checks the format of a given CSRF token not if it has been tempered with.
+validFormatCsrfToken :: CsrfToken -> Bool
+validFormatCsrfToken (CsrfToken st)
+    | Right s' <- convertFromBase Base16 (cs st :: SBS) = SBS.length s' == 64
+    | otherwise                                         = False
+
+-- | Computes a valid CSRF token given a nonce.
+makeCsrfToken :: CsrfNonce -> FAction CsrfToken
+makeCsrfToken (CsrfNonce rnd) = do
+    maySessionToken <- preuse $ fsdLogin . _Just . fslToken
+    case maySessionToken of
+        Nothing -> crash $ FActionError500 "No session token"
+        Just sessionToken -> do
+            CsrfSecret key <- getCsrfSecret
+            return $ CsrfToken . cs $ rnd <> convertToBase Base16 (hmac key (tok <> rnd) :: HMAC SHA256)
+          where
+            tok = cs $ fromThentosSessionToken sessionToken
+
+-- | Extracts the nonce part from the CSRF token.
+csrfNonceFromCsrfToken :: CsrfToken -> CsrfNonce
+csrfNonceFromCsrfToken = CsrfNonce . SBS.take 64 . cs . fromCsrfToken
+
+-- | Checks given a 'CsrfToken'.
+-- This token should from the form data of the POST request.
```

Comment:
Indeed...

Verify the authenticity of a given 'CsrfToken'. This token should come from the form data of the POST request NOT the one from the 'FrontendSessionData'.